### PR TITLE
Spark move with history

### PR DESCRIPTION
**Pull request by OleksandrDvornik**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -54,6 +54,31 @@ jobs:
...
+    working_directory: ~/openlineage/integration/spark
+    docker:
+      - image: circleci/openjdk:8-jdk
+    steps:
+      - checkout
+      - restore_cache:
+          keys:
+            - v1-integration-spark-{{ .Branch }}-{{ .Revision }}
+            - v1-integration-spark-{{ .Branch }}
+      - run: ./integration/spark/gradlew -p integration/spark/ --no-daemon --stacktrace build
```
Comment (wslulciuc) [inline .circleci/config.yml:67]: Do you need to prefix the gradle cmds with `./integration/spark` if the CI working directly is `~/openlineage/integration/spark`? All cmds should be relative to the working dir.

File: .circleci/config.yml [hunk 2]
```diff
@@ -79,6 +104,35 @@ jobs:
             - ./dist/*.whl
             - ./dist/*.tar.gz
 
+  integration-test-spark:
+    working_directory: ~/openlineage/integration/spark
+    machine: true
+    environment:
+      TESTCONTAINERS_RYUK_DISABLED: "true"
+      JDK8_HOME: /usr/lib/jvm/java-8-openjdk-amd64
```
Comment (wslulciuc) [inline .circleci/config.yml:112]: We've been meaning to change `JDK8_HOME` env var to `JAVA8_HOME`. Mind renaming as part of this PR?

Comment (wslulciuc) [reply .circleci/config.yml:112]: Actually, let's just leave it as is to avoid breaking dev environments.

File: .circleci/config.yml [hunk 3]
```diff
@@ -97,6 +151,8 @@ workflows:
           app_src_directory: client/java
       - test-client-python
       - test-integration-common
+      - build-integration-spark
+      - integration-test-spark
```
Comment (wslulciuc) [inline .circleci/config.yml:155]: The `integration-test-spark` CI job should depend on `build-integration-spark`, see https://github.com/MarquezProject/marquez/blob/main/.circleci/config.yml#L339

File: integration/spark/README.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+Create a new notebook and paste the following into the first cell:
+```python
+from pyspark.sql import SparkSession
+
+spark = (SparkSession.builder.master('local')
+         .appName('sample_spark')
+         .config('spark.jars.packages', 'io.github.openlineage:openlineage-spark-0.1-SNAPSHOT.jar'
+         .config('spark.extraListeners', 'openlineage.spark.agent.OpenLineageSparkListener')
+         .config('openlineage.url', 'http://marquez:5000/api/v1/namespaces/spark_integration/')
```
Comment (wslulciuc) [inline integration/spark/README.md:55]: The config `openlineage.url` should have the `spark.*` prefix

File: integration/spark/src/main/java/openlineage/spark/agent/OpenLineageSparkListener.java [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+import org.apache.spark.scheduler.SparkListenerApplicationStart;
+import org.apache.spark.scheduler.SparkListenerEvent;
+import org.apache.spark.scheduler.SparkListenerJobEnd;
+import org.apache.spark.scheduler.SparkListenerJobStart;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.execution.ui.SparkListenerSQLExecutionEnd;
+import org.apache.spark.sql.execution.ui.SparkListenerSQLExecutionStart;
+
+@Slf4j
+public class OpenLineageSparkListener extends org.apache.spark.scheduler.SparkListener {
```
Comment (collado-mike) [inline integration/spark/src/main/java/openlineage/spark/agent/OpenLineageSparkListener.java:42]: You can import the `SparkListener` class, now that ours has a distinct name.

File: integration/spark/src/main/java/openlineage/spark/agent/client/OpenLineageClient.java [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.apache.hc.client5.http.async.methods.BasicHttpRequests;
+import org.apache.hc.client5.http.impl.async.CloseableHttpAsyncClient;
+import org.apache.hc.client5.http.impl.async.HttpAsyncClients;
+import org.apache.hc.core5.http.HttpRequest;
+import org.apache.hc.core5.http.HttpResponse;
+import org.apache.hc.core5.http.Message;
+
+@Slf4j
+public class OpenLineageClient {
```
Comment (collado-mike) [inline integration/spark/src/main/java/openlineage/spark/agent/client/OpenLineageClient.java:34]: Let's go with `OpenLineageHttpClient`, as this is HTTP specific

File: integration/spark/src/main/java/openlineage/spark/agent/lifecycle/LogicalPlanTraverser.java [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+import org.apache.spark.sql.execution.datasources.InsertIntoHadoopFsRelationCommand;
+import org.apache.spark.sql.execution.datasources.LogicalRelation;
+import org.apache.spark.sql.execution.datasources.jdbc.JDBCOptions;
+import org.apache.spark.sql.execution.datasources.jdbc.JDBCRelation;
+import org.apache.spark.sql.sources.BaseRelation;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+
+public abstract class LogicalPlanTraverser {
```
Comment (collado-mike) [inline integration/spark/src/main/java/openlineage/spark/agent/lifecycle/LogicalPlanTraverser.java:28]: I don't think this is used anymore. Let's just delete it.

File: integration/spark/src/test/java/openlineage/spark/agent/SparkContainerIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+
+  @BeforeAll
+  public static void setup() {
+    mockServerClient =
+        new MockServerClient(
+            openLineageClientMockContainer.getHost(),
+            openLineageClientMockContainer.getServerPort());
+    mockServerClient
+        .when(request("/api/v1/lineage"))
+        .respond(org.mockserver.model.HttpResponse.response().withStatusCode(200));
```
Comment (collado-mike) [inline integration/spark/src/test/java/openlineage/spark/agent/SparkContainerIntegrationTest.java:53]: I think it ought to be a `201`, right?

File: integration/spark/src/test/java/openlineage/spark/agent/SparkContainerIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+        .withStartupTimeout(Duration.of(2, ChronoUnit.MINUTES))
+        .withEnv(
+            ImmutableMap.of(
+                "POSTGRES_USER", "postgres",
+                "POSTGRES_PASSWORD", "password",
+                "MARQUEZ_DB", "marquez",
+                "MARQUEZ_USER", "marquez",
+                "MARQUEZ_PASSWORD", "marquez"))
+        .withFileSystemBind("./docker/init-db.sh", "/docker-entrypoint-initdb.d/init-db.sh");
+  }
```
Comment (collado-mike) [inline integration/spark/src/test/java/openlineage/spark/agent/SparkContainerIntegrationTest.java:99]: I don't think you need the postgres container anymore, right?

File: integration/spark/README.md [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+```
+
+## Getting started
+
+### Quickstart
+The fastest way to get started testing Spark and OpenLineage is to use the docker-compose files included
+in the project. From the spark integration directory ($OPENLINEAGE_ROOT/integrations/spark) execute
+```bash
+docker-compose up
+```
```
Comment (collado-mike) [inline integration/spark/README.md:30]: This is no longer accurate. There's no local marquez host configured in the docker-compose file- maybe we can just add an instance of marquez in the docker-compose file for the sake of the quickstart?

File: integration/spark/src/test/resources/org/apache/spark/log4j-defaults.properties [hunk 11]
```diff
@@ -0,0 +1,9 @@
+# Set everything to be logged to the console
+log4j.rootCategory=INFO, console
+log4j.appender.console=org.apache.log4j.ConsoleAppender
+log4j.appender.console.target=System.err
+log4j.appender.console.layout=org.apache.log4j.PatternLayout
+log4j.appender.console.layout.ConversionPattern=%d{yy/MM/dd HH:mm:ss} %p %c{1}: %m%n
+
+# set the log level for the marquez library
```
Comment (collado-mike) [inline integration/spark/src/test/resources/org/apache/spark/log4j-defaults.properties:8]: marquez->openlineage

**Review by collado-mike — COMMENTED** [review]
Very nice. I like the mockserver approach. I think we can use this for all the integration tests in the migration.

File: .circleci/config.yml (hunk 1 above)
Comment (OleksandrDvornik) [reply .circleci/config.yml:67]: As I got working dir, is where the code should be checkout, it's my view from my tests. I'll double check that

File: .circleci/config.yml (hunk 3 above)
Comment (OleksandrDvornik) [reply .circleci/config.yml:155]: You mean that we don't need to run Integration tests if build failed? Cause it's not dependent directly, we can run ITs without build, task run shadowJar and then run tests. I'll change it in the same way as it was

File: integration/spark/README.md (hunk 10 above)
Comment (OleksandrDvornik) [reply integration/spark/README.md:30]: It's good idea as for me, will add it.

File: integration/spark/README.md (hunk 4 above)
Comment (OleksandrDvornik) [reply integration/spark/README.md:55]: I guess not, it's the last config which wasn't wrapped with "spark." prefix. 
[marquez](https://github.com/MarquezProject/marquez/blob/main/integrations/spark/src/main/java/marquez/spark/agent/SparkListener.java#L262)
I was thinking to change that, but maybe in next PR, don't want to mix things here

File: integration/spark/README.md [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+         .config('spark.sql.repl.eagerEval.enabled', 'true')
+         .getOrCreate())
+```
+When running on a real cluster, the openlineage-spark jar has to be in a known location on the master
+node of the cluster and its location referenced in the `spark.driver.extraJavaOptions` parameter.
+## Arguments
+
+### Spark Listener
+The SparkListener reads its configuration from SparkConf parameters. These can be specified on the
+command line (e.g., `--conf "spark.openlineage.url=http://marquez:5000/api/v1/namespaces/my_namespace/job/the_job"`)
```
[the remark this answers is not in the document]
Comment (OleksandrDvornik) [reply integration/spark/README.md:101]: not sure, would work "spark.openlineage.url"?I'll check that

File: integration/spark/README.md [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+The java agent accepts an argument in the form of a uri. It includes the location of Marquez, the
+namespace name, the parent job name, and a parent run id. The run id will be emitted as a parent run
+facet.
+```
+{marquez_home}/api/v1/namespaces/{namespace}/job/{job_name}/runs/{run_uuid}?api_key={api_key}"
+
+```
+For example:
+```
+https://marquez.example.com:5000/api/v1/namespaces/foo/job/spark.submit_job/runs/a95858ad-f9b5-46d7-8f1c-ca9f58f68978"
```
[the remark this answers is not in the document]
Comment (OleksandrDvornik) [reply integration/spark/README.md:121]: done

File: integration/spark/README.md [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+```
+
+# Compatibility Notes
+Tested and compatible for Spark `2.4.7` only.
+
```
[the remark this answers is not in the document]
Comment (OleksandrDvornik) [reply integration/spark/README.md:126]: done

File: integration/spark/src/test/java/openlineage/spark/agent/SparkContainerIntegrationTest.java (hunk 8 above)
Comment (OleksandrDvornik) [reply integration/spark/src/test/java/openlineage/spark/agent/SparkContainerIntegrationTest.java:53]: You are correct should be 201, we  have POST there, thanks

File: integration/spark/README.md (hunk 4 above)
Comment (OleksandrDvornik) [reply integration/spark/README.md:55]: so, it's not working with spark.openlineage.url.
```
 "--conf",
 "spark.openlineage.url=" + "http://openlineageclient:1080/api/v1",
```
I've changed SparkContainerIntegrationTest to use url, and it fails with 
` org.apache.hc.core5.http.ProtocolException: Target host is not specified)`
So, I guess we should fix that

File: .circleci/config.yml (hunk 1 above)
Comment (OleksandrDvornik) [reply .circleci/config.yml]: I've changed  -checkout step to - *checkout_project_root, and ./gradlew works fine without prefixes

**Review by collado-mike — APPROVED** [review]

File: integration/spark/src/main/java/openlineage/spark/agent/OpenLineageSparkContext.java [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.concurrent.ForkJoinPool;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import openlineage.spark.agent.client.LineageEvent;
+import openlineage.spark.agent.client.OpenLineageClient;
+import openlineage.spark.agent.client.OpenLineageHttpException;
+import openlineage.spark.agent.client.ResponseMessage;
+
+@Slf4j
+public class OpenLineageSparkContext {
```
[the remark this answers is not in the document]
Comment (wslulciuc) [reply integration/spark/src/main/java/openlineage/spark/agent/OpenLineageSparkContext.java:16]: @collado-mike We can rename after the merge, https://github.com/OpenLineage/OpenLineage/issues/100

File: integration/spark/src/main/java/openlineage/spark/agent/client/OpenLineageClient.java (hunk 6 above)
Comment (wslulciuc) [reply integration/spark/src/main/java/openlineage/spark/agent/client/OpenLineageClient.java:34]: I think this should be replaced by the `openlineage-java` client?

**Review by wslulciuc — APPROVED** [review]
Great work, @OleksandrDvornik! We'll want to do some follow up work on the spark integration, but overall good to merge 💯 🥇

**Pull request merged**